### PR TITLE
Test harness

### LINKED ISSUE #53: Test harness
Currently all testing facilities are disabled for the crates in this repo. It makes sense because everything will only run on an actual Flipper Zero, but it does make verification of things tricky.

It would be useful to have a test harness, that runs the test binaries on a connected Flipper Zero. I'm imagining something like the [integration tests for the `yubikey` crate](https://github.com/iqlusioninc/yubikey.rs/blob/main/tests/integration.rs), which require setting a `YUBIKEY_SERIAL` environment variable in order to run. The [`test-with`](https://crates.io/crates/test-with) crate might also be helpful here, specifically `#[test_with::env(FLIPPER_ZERO_SERIAL_PORT)]` or something, that can be used to enable at compilation time tests that require a connected Flipper Zero.

### PR DESCRIPTION
**Pull request by str4d**
Closes #53.

### CODE REVIEW

File: crates/.cargo/config.toml
Comment (str4d) [inline crates/.cargo/config.toml:2]: This works reliably from inside the `crates/` directory, but I have not tested from elsewhere yet.

File: crates/flipperzero/src/lib.rs
Comment (str4d) [inline crates/flipperzero/src/lib.rs:21]: This infrastructure will likely end up becoming a proc macro (so it can be used for both unit tests and integration tests), but for now I'm implementing it explicitly to make its actions clear.

File: tools/src/bin/run-fap.rs
Comment (str4d) [inline tools/src/bin/run-fap.rs:95]: Currently it is not possible to provide arguments to external FAPs. I've opened https://github.com/flipperdevices/flipperzero-firmware/issues/2505 upstream requesting this. In the meantime, this just means that test filtering won't work.

File: crates/flipperzero/src/lib.rs
Comment (str4d) [reply crates/flipperzero/src/lib.rs]: I've now made it a proc macro, because it ended up being easier (as then I could use a second proc macro to process the tests themselves into a form that the test runner can use).

File: crates/test/macros/src/lib.rs
Comment (str4d) [inline crates/test/macros/src/lib.rs:165]: Currently the test runner prints the same information as the [`flipperzero-firmware` unit tests](https://github.com/flipperdevices/flipperzero-firmware/blob/0.78.1/documentation/UnitTests.md) in the device logs. However, for the purpose of `cargo test` we need to extract this information back to the host. I currently believe this means we'll need to save the results to a file and then download it before cleaning up the device.

File: crates/test/macros/src/lib.rs
Comment (str4d) [inline crates/test/macros/src/lib.rs:141]: The current state of this PR does successfully run tests on the Flipper Zero. However, if any test assertion fails, the unit test FAP crashes so we won't get any results. Given how `flipperzero_rt` intercepts the panic handler, and `catch_unwind` isn't available anyway without `std`, I think the best solution is to require tests to return `Result<(), &'static str>`, and then maybe use the proc macro to rewrite the assertion-based tests if possible.

Comment (str4d) [reply crates/test/macros/src/lib.rs]: Intercepting the `assert` macros works very nicely!

File: crates/test/macros/src/deassert.rs
Comment (str4d) [inline crates/test/macros/src/deassert.rs:9]: In addition to these macros, we should also try to hunt for and replace `.unwrap()` and `.expect(..)` with something that similarly causes `Err(..)` to be returned. That being said, I'm fine with this being best-effort; the goal is to make writing unit tests easy, not to completely erase knowledge of the fact that panics will break `cargo test`.

File: tools/src/bin/run-fap.rs
Comment (str4d) [inline tools/src/bin/run-fap.rs:82]: I haven't yet figured out if any of the Flipper Zero SDKs let a running external FAP learn where it was loaded from. Without that knowledge, it's gonna be tricky to reliably redirect stdout to a file that both won't collide with an existing file on the Flipper Zero, and that this tool can find.

**Comment by str4d** [comment]
Okay, I have this working on Linux and macOS! Current output:
```
$ cd crates/
$ cargo test
   Compiling proc-macro2 v1.0.52
   Compiling unicode-ident v1.0.8
   Compiling quote v1.0.26
   Compiling syn v1.0.109
   Compiling ufmt-write v0.1.0
   Compiling flipperzero-rt v0.7.2 (/path/to/flipperzero/crates/rt)
   Compiling ufmt-macros v0.3.0
   Compiling flipperzero-test-macros v0.1.0 (/path/to/flipperzero/crates/test/macros)
   Compiling ufmt v0.2.0
   Compiling flipperzero-sys v0.7.2 (/path/to/flipperzero/crates/sys)
   Compiling flipperzero-test v0.1.0 (/path/to/flipperzero/crates/test)
   Compiling flipperzero-alloc v0.7.2 (/path/to/flipperzero/crates/alloc)
   Compiling flipperzero v0.7.2 (/path/to/flipperzero/crates/flipperzero)
    Finished test [optimized + debuginfo] target(s) in 4.73s
     Running unittests src/lib.rs (target/thumbv7em-none-eabihf/debug/deps/flipperzero-ec498257ef00abdb)

running 3 tests
test dolphin::tests::stats ... ok
test furi::message_queue::tests::capacity ... ok
test furi::sync::tests::unshared_mutex_does_not_block ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2ms
leaked: 88 bytes

$ # Unplug Flipper Zero
$ cargo test
    Finished test [optimized + debuginfo] target(s) in 0.01s
     Running unittests src/lib.rs (target/thumbv7em-none-eabihf/debug/deps/flipperzero-ec498257ef00abdb)
Error: unable to find Flipper Zero
error: test failed, to rerun pass `-p flipperzero --lib`

Caused by:
  process didn't exit successfully: `/path/to/flipperzero/crates/../tools/cargo-runner.sh /path/to/flipperzero/crates/target/thumbv7em-none-eabihf/debug/deps/flipperzero-ec498257ef00abdb` (exit status: 1)
```

File: tools/src/bin/run-fap.rs
Comment (str4d) [reply tools/src/bin/run-fap.rs:88]: I ended up just using a common fixed path of `/ext/flipperzero-rs-stdout`, on the assumption that external FAPs cannot run concurrently, so only one app at a time will ever be running, and as long as we immediately read the file and then remove it we're probably fine. This will do until the Flipper Zero SDK supports RPC usage.

File: crates/test/macros/src/lib.rs
Comment (str4d) [reply crates/test/macros/src/lib.rs]: Saving the results to a file is what I am doing now.

**Comment by str4d** [comment]
And if I manually cause some tests to fail:
```
$ cargo test
   Compiling flipperzero v0.7.2 (/path/to/flipperzero/crates/flipperzero)
    Finished test [optimized + debuginfo] target(s) in 0.28s
     Running unittests src/lib.rs (target/thumbv7em-none-eabihf/debug/deps/flipperzero-ad4e99e6ec932a9a)

running 3 tests
test dolphin::tests::stats ... ok
test furi::message_queue::tests::capacity ... FAILED

---- furi::message_queue::tests::capacity stdout ----
assertion failed: queue.len() == 1

test furi::sync::tests::unshared_mutex_does_not_block ... FAILED

---- furi::sync::tests::unshared_mutex_does_not_block stdout ----
assertion failed: * value == 6


test result: FAILED. 1 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 5ms
leaked: 88 bytes
```

Currently the test failures are printed in-line, to avoid requiring `alloc`.

File: tools/src/bin/run-fap.rs
Comment (JarvisCraft) [reply tools/src/bin/run-fap.rs:88]: You may try addressing the issues in Flipperzero [Telegram group](https://t.me/flipperzero), i.e. `Development (SW&HW)` channel.

Comment (str4d) [reply tools/src/bin/run-fap.rs:88]: https://github.com/flipperdevices/flipperzero-firmware/issues/2505#issuecomment-1474184238 indicated that an improved host-device communication protocol is planned for external FAPs, so that is likely what will be used instead when available.

**Comment by str4d** [comment]
I've adjusted the test harness so it now also works for integration tests, and to allow the manifest arguments to be set (in particular, while adding more tests in subsequent local changes I need to bump the stack size).

There are probably more usability tweaks that could be made, but I think we'll discover those as the test harness is used, so I'm not planning to make any more changes to this PR other than in response to review.

File: crates/test/macros/src/deassert.rs
Comment (dcoles) [reply crates/test/macros/src/deassert.rs:9]: That seems like a reasonable approach. Since we don't support `panic = "unwind"` (and [likely never will](https://github.com/flipperdevices/flipperzero-firmware/issues/1851)), I don't really see any other way of running tests than a) 1 binary per test, b) co-operative tests like you're doing here.

File: crates/test/macros/src/deassert.rs
Comment (dcoles) [inline crates/test/macros/src/deassert.rs:96]: What does this function do? It looks like it might be for unpacking a 2-argument macro into individual expressions.

**Review by dcoles — APPROVED** [review]

File: crates/.cargo/config.toml
Comment (str4d) [reply crates/.cargo/config.toml:2]: It turns out that `target.<cfg>.runner` also works, so I might be able to specify a Windows script here as well with some effort.

**Comment by str4d** [comment]
[Rebased on `main`](https://github.com/flipperzero-rs/flipperzero/compare/754d827f14ad03401bf0a8bdeaa0e9455439aab7..91452b59aca13dcc2f2b5802cd410413cfab23a4) to fix merge conflict with #59.

File: crates/.cargo/config.toml
Comment (str4d) [reply crates/.cargo/config.toml:2]: I can't figure out an easy way to get this to work, so given that Unix platforms almost always have Python, and you're clearly already requiring it on Windows, let's just use that for now.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/flipperzero-rs/flipperzero/compare/91452b59aca13dcc2f2b5802cd410413cfab23a4..bab2e58543afd8d32c10081eaf576c0e803a9731) to replace `cargo-runner.sh` with `cargo-runner.py`. I've tested that this PR still works on Ubuntu and macOS; it should now also work on Windows (untested, but [Windows can execute `*.py` files directly if their extension is associated correctly](https://docs.python.org/3/faq/windows.html#how-do-i-make-python-scripts-executable)

File: crates/cargo-runner.py
Comment (dcoles) [inline crates/cargo-runner.py:22]: Why are you printing a newline here?

File: crates/cargo-runner.py
Comment (dcoles) [inline crates/cargo-runner.py:35]: For something as simple as this script, we could probably even get away with just `+ sys.argv`.

**Comment by dcoles** [comment]
@str4d It looks like the runner [might be causing problems for GitHub Actions](https://github.com/flipperzero-rs/flipperzero/actions/runs/4581112299/jobs/8090398886?pr=57). Can you take a look?

**Comment by dcoles** [comment]
> it should now also work on Windows (untested, but [Windows can execute *.py files directly if their extension is associated correctly](https://docs.python.org/3/faq/windows.html#how-do-i-make-python-scripts-executable)).

Looks like `runner` must be an application, not just an associated file type:

```
PS> ./cargo-runner.py
usage: cargo-runner.py [-h] binary ...
cargo-runner.py: error: the following arguments are required: binary, arguments
PS> cargo test
    Finished test [optimized + debuginfo] target(s) in 0.03s
     Running unittests src\lib.rs (target\thumbv7em-none-eabihf\debug\deps\flipperzero-b5d0b378c3208e7a)
error: test failed, to rerun pass `-p flipperzero --lib`

Caused by:
  could not execute process `C:\Users\dcoles\workspace\flipperzero-rs\crates\./cargo-runner.py C:\Users\dcoles\workspace\flipperzero-rs\crates\target\thumbv7em-none-eabihf\debug\deps\flipperzero-b5d0b378c3208e7a` (never executed)

Caused by:
  %1 is not a valid Win32 application. (os error 193)
```

In this case, maybe we just go with the shell script and make this work for Windows in a future PR.

File: crates/cargo-runner.py
Comment (str4d) [reply crates/cargo-runner.py:22]: Oops, that's left over from debugging.

**Comment by str4d** [comment]
> > it should now also work on Windows (untested, but [Windows can execute *.py files directly if their extension is associated correctly](https://docs.python.org/3/faq/windows.html#how-do-i-make-python-scripts-executable)).
> 
> Looks like `runner` must be an application, not just an associated file type:

I've now found https://github.com/rust-lang/cargo/issues/8347 which indicates we just need to set the runner to `python path/to/script.py`. I'll update the PR to do that, and then you can retest on Windows.

**Comment by str4d** [comment]
> @str4d It looks like the runner [might be causing problems for GitHub Actions](https://github.com/flipperzero-rs/flipperzero/actions/runs/4581112299/jobs/8090398886?pr=57). Can you take a look?

I think what's going on here is that CI was configured to run all tests, but nothing had tests enabled so it did nothing. Now that we enable tests, CI tries to run them, but because the `.cargo/config` sets the default target to be not the host, they fail. I think the solution is to just not run any tests in CI, as there are no useful tests to be run outside the Flipper Zero.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/flipperzero-rs/flipperzero/compare/bab2e58543afd8d32c10081eaf576c0e803a9731..e869e356a17de0b18d961b2f4b2e72f6da83e028) to add missing package to CI, and hopefully fix `cargo test` on Windows.

**Comment by str4d** [comment]
Okay, confirmed that CI fails now because it can't find a Flipper Zero. I'll adjust CI to expect that failure, so we're still compiling everything and checking the runner.

**Comment by dcoles** [comment]
> I've now found https://github.com/rust-lang/cargo/issues/8347 which indicates we just need to set the runner to python path/to/script.py. I'll update the PR to do that, and then you can retest on Windows.

Looks like I hit python/cpython#99185 (Windows does not provide a `python3.exe`), which is totally dumb and bites me all the time. It would be possible to make the script compatible with both Python 2 and 3, but on my Ubuntu box there is no `python` binary unless I install the `python-is-python3` package.

[PEP-397](https://peps.python.org/pep-0397) tries to resolve this situation by introducing a `py.exe` that can interpret the shbang line, but this isn't available on Ubuntu (not to be confused with [`pythonpy`](https://packages.ubuntu.com/jammy/pythonpy)).

Not sure if there's an easy fix. There doesn't seem to be a `cfg(host_os = "windows")` and I can't figure out a way to get `cmd.exe` and `bash` to accept the same `runner` line.

**Comment by str4d** [comment]
Hmm. Now that I know we can have arguments inside `runner`, I wonder if we can just call `cargo` directly instead of using a wrapper script? The purpose of the script was to encapsulate the numerous arguments, but we also know that `cargo` should be available on the `PATH`. I'll give this a go.

**Comment by str4d** [comment]
> The purpose of the script was to both encapsulate the numerous arguments and enable changing working directory so that the `.cargo/config` won't apply. If the latter can be worked around, then this might work.

I cannot figure out how to work around the latter inside `target.<target>.runner` without adding a `cd ... &&`, which is not cross-platform. However, reading a bunch of issues (around trying to unset `build.target` (https://github.com/rust-lang/cargo/issues/8112) or reset it to the host (https://github.com/rust-lang/cargo/issues/8687)) led me to discover [`per-package-target`](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#per-package-target), which seems like it would also resolve the issue. We would set each of the crates under `crates/` to have a `package.forced-target`, and then remove the `build.target` from `crates/.cargo/config.toml`. This would mean that calling `cargo run --manifest-path ../tools/Cargo.toml` from inside `crates/` would then compile for the host instead of the Flipper Zero. However, it would block #58 on another unstable feature.

**Comment by str4d** [comment]
Sadly `per-package-target` doesn't work, because even though `forced-target` correctly causes the test binary to be compiled for the Flipper Zero, it doesn't cause the corresponding per-target runner to be used (so instead the binary runs directly on the host). I've asked about this in the tracking issue (https://github.com/rust-lang/cargo/issues/9406#issuecomment-1493318519).

However, I've figured out another solution: aliases! We remove the `build.target` (so by default we build for the host), and then add aliases that set the correct `--target`. I've pushed a commit that does this, providing `*-flipper` aliases for `check`, `build`, and `test` (happy to rename to something else / shorter if preferred). This does interfere with muscle memory though (so it's a trade-off between "need to remember separate commands" with this commit, and "need to make custom changes to my OS environment to get Python working" before), and it means that e.g. the `rust-analyzer` plugin for VSCode needs to be configured to use the correct target (I can commit that if you want).

**Comment by str4d** [comment]
@dcoles is there anything else I need to do before this can be merged? I'd love to get back on with my other PRs like #50 that are blocked on testing.

**Comment by str4d** [comment]
> Over the weekend I got to see how a few other projects tackle this. Ferrus Systems has a tool called [`probe-run`](https://crates.io/crates/probe-run) that can be used as a runner for a number of embedded platforms (though probably not compatible with Flipper apps). You can use the [`hf2-cli`](https://crates.io/crates/hf2-cli) in a similar way via `hf2 elf`.

Yeah, those were what made me realize we could run tests and apps in this way. But we can't use any of the existing ones because they flash firmware, and I wanted this to be as easy to use as possible, so no need to install anything ideally. Didn't quite get there on all platforms, and if we end up having an installable FAP runner then I don't mind.

**Comment by str4d** [comment]
The other thing is that an installable FAP runner will be much nicer once upstream has implement proper RPC support for FAPs, so we don't have to do the hacky "write stdout to a file and then copy it off afterwards" that really is only suitable for one-shot FAPs.

**Comment by dcoles** [comment]
> The other thing is that an installable FAP runner will be much nicer once upstream has implement proper RPC support for FAPs, so we don't have to do the hacky "write stdout to a file and then copy it off afterwards" that really is only suitable for one-shot FAPs.

I'm very much looking forward to that. It's bit awful what you have to do when trying to use a serial CLI as a file transfer layer.

**Pull request merged**
